base/linalg/lq.jl

```diff
@@ -90,19 +88,25 @@ convert(::Type{LQPackedQ{T}}, Q::LQPackedQ) where {T} = LQPackedQ(convert(Abstra
 convert(::Type{AbstractMatrix{T}}, Q::LQPackedQ) where {T} = convert(LQPackedQ{T}, Q)
 convert(::Type{Matrix}, A::LQPackedQ) = LAPACK.orglq!(copy(A.factors),A.τ)
 convert(::Type{Array}, A::LQPackedQ) = convert(Matrix, A)
-function full(A::LQPackedQ{T}; thin::Bool = true) where T
-    #= We construct the full eye here, even though it seems inefficient, because
-    every element in the output matrix is a function of all the elements of
-    the input matrix. The eye is modified by the elementary reflectors held
-    in A, so this is not just an indexing operation. Note that in general
-    explicitly constructing Q, rather than using the ldiv or mult methods,
-    may be a wasteful allocation. =#
-    if thin
-        convert(Array, A)
-    else
-        A_mul_B!(A, eye(T, size(A.factors,2), size(A.factors,1)))
-    end
-end
+
+"""
+    squareQ(Q::LQPackedQ) -> Matrix
+
+Construct an explicit, square form of `Q`.
+
+!!! note
+    Constructing a explicit representation of `Q`, rather than using, e.g., `ldiv` or `mult`
+    methods with the implicit, packed representation of `Q`, may be less efficient and/or
+    require unnecessary allocations.
+
+!!! note "Implementation"
+    This method initially constructs a dense identity matrix, even though that may seem
+    inefficient, because every element in the returned matrix is a function of every element
+    in the (implicit) input matrix `Q`. That identity matrix is modified by the elementary
+    reflectors held in `Q`, so this operation does not reduce to simple indexing.
+"""
+squareQ(Q::LQPackedQ) = A_mul_B!(Q, eye(eltype(Q), reverse(size(Q.factors))...))
```

Comment:
This is not square. Should probably be `eye(eltype(Q), size(Q.factors, 2)).`

Reply:
Good catch! I failed to read the existing code critically. And agreed, `eye(eltype(Q), size(Q.factors, 2))` looks correct. What was the object the original code returns?

Reply:
It looks like this PR is turning into a significant bug fix. The current version gives
```julia
julia> full(lqfact(randn(4,2))[:Q], thin = true)
2×2 Array{Float64,2}:
 -0.759173  0.650889
  0.650889  0.759173

julia> full(lqfact(randn(4,2))[:Q], thin = false)
2×4 Array{Float64,2}:
 -0.652378  -0.757894  0.0  0.0
 -0.757894   0.652378  0.0  0.0

julia> full(lqfact(randn(2,4))[:Q], thin = false)
4×2 Array{Float64,2}:
 -0.22708   -0.390319
  0.72372   -0.204318
 -0.025446   0.888878
 -0.651166  -0.125704

julia> full(lqfact(randn(2,4))[:Q], thin = false)
4×2 Array{Float64,2}:
 -0.0604746   0.111308
  0.822683   -0.406272
 -0.158067   -0.81894
  0.542725    0.389731
```
so only the first one is correct. Here is how they are supposed to look like

![2017-09-18 09 25 21](https://user-images.githubusercontent.com/505001/30532484-cf60e2fe-9c54-11e7-89e4-748239e27279.jpg)